[Dist] Add fallback reduce_scatter_base, allgather_base APIs to Gloo

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112145
* __->__ #112144

Per Ke's suggestion, adding these APIs in ProcessGroupGloo directly to
enable FSDP on CPUs

Differential Revision: [D50636382](https://our.internmc.facebook.com/intern/diff/D50636382/)

rohan-varma committed:
[Dist] Add fallback reduce_scatter_base, allgather_base APIs to Gloo

Per Ke's suggestion, adding these APIs in ProcessGroupGloo directly to
enable FSDP on CPUs

Differential Revision: [D50636382](https://our.internmc.facebook.com/intern/diff/D50636382/)

[ghstack-poisoned]

wz337 reviewed (approved):
Approved via phabricator.

fegin reviewed (approved):
Should we have some unit tests for these?

wanchaol commented on torch/csrc/distributed/c10d/ProcessGroupGloo.cpp:
```diff
@@ -1919,6 +1919,38 @@ class AsyncAllgatherCUDAWork : public AsyncAllgatherWork {
 
 } // namespace
 
+c10::intrusive_ptr<Work> ProcessGroupGloo::_reduce_scatter_base(
```
Nice, maybe worth removing those fallbacks in the functional collective paths and this might help testing https://github.com/pytorch/pytorch/blob/main/torch/distributed/_functional_collectives_impl.py#L180

wanchaol reviewed (approved):
lgtm

rohan-varma committed:
Update on "[Dist] Add fallback reduce_scatter_base, allgather_base APIs to Gloo"


Per Ke's suggestion, adding these APIs in ProcessGroupGloo directly to
enable FSDP on CPUs

Differential Revision: [D50636382](https://our.internmc.facebook.com/intern/diff/D50636382/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Dist] Add fallback reduce_scatter_base, allgather_base APIs to Gloo"


Per Ke's suggestion, adding these APIs in ProcessGroupGloo directly to
enable FSDP on CPUs

Differential Revision: [D50636382](https://our.internmc.facebook.com/intern/diff/D50636382/)

[ghstack-poisoned]
